Write, storeInBuff, emptyBuff macros

miporto opened the pull request:
!!!!! registers used when calling writing, storeInBuff macros must be changed into ones that are available !!!!! Check lines ending with comment " # CHANGE INTO ... "

miporto committed:
Add store in buffer and empty buffer macros

miporto committed:
Use storeInBuff for storing pgm header

miporto committed:
Add writem calls but keep old ones for now

miporto commented:
@saantiaguilera @abrden

Manuporto committed:
Correct indentation

saantiaguilera commented on julia-assembly/src/mips32_plot.S:
```diff
@@ -27,6 +29,16 @@ error_msg:
+# Going to need a register to store the buff location
+.macro storeInBuff (%data, %len)
+    	sw $data, ($t0) # CHANGE INTO PROPER REG
+    	addi $t0, $zero, %len
```
`addi $t0, 0, $len` ?? (y tambien t0 tendria que ser el registro del buffer :P)

saantiaguilera replied:
Lo ideal no seria que nosotros solo usemos storeInBuff ?? Y esta macro llame a writem si se lleno el buffer (hacemos posBuffer - posBufferIni para saber el size) y despues a emptyBuff para resetearlo.

Tambien vamos a tener que llamar a fflush cuando terminamos de iterar en todos los pixels para vaciar lo que no se llego a llenar.

miporto replied:
> `addi $t0, 0, $len` ?? (y tambien t0 tendria que ser el registro del buffer :P) 

$t0 tendria que ser un registro reservado para esto especificamente. Lo que hago en esa instruccion es sumarle la longitud (en bytes) de lo que se acaba de almacenar

saantiaguilera commented on julia-assembly/src/mips32_plot.S:
```diff
@@ -52,6 +64,18 @@ mips32_plot:
     # Write PGM header
 
+    lw $t4, (header)
+    storeInBuff $t1 3
```
Porque t1 ? Si guardaste la word en t4 (talves estoy medio bobo y me perdi algo jaja)

miporto replied:
Tenes razon, es un typo

saantiaguilera commented on julia-assembly/src/mips32_plot.S:
```diff
@@ -184,12 +214,21 @@ UL_im_icr:
+fflush:
+    #TODO it flushes remaining data into file, but if thats the case it results
+    # in an error
+    lw $t1, buffer_ini # CHANGE INTO PROPER REG
```
Y flush yo haria que solo sea un writem -> emptyBuf (chequeando antes de que el size > 0 del buffer)

miporto replied:
Claro pero en el codigo te tira un error si el resultado del flush es distinto de 0

miporto replied:
En la linea 224 se hace todo eso (fflush_error)

saantiaguilera committed:
Indents

saantiaguilera committed:
Typo

saantiaguilera committed:
Fix typo

saantiaguilera committed:
Checking buffer length and writing when needed

saantiaguilera committed:
Updated fflush

saantiaguilera commented:
Sigo viendo medio raro algunas cosas :P ahi hice los cambios que deciamos (y lo del buffer del wpp que hablamos)

Manuporto committed:
Add fd as an argument when using writem macro

Manuporto committed:
Change macro call into label and then call the macro there

We need to store the sp location before the branching so the after the writing its possible to jump back to the next instruction of the macro

saantiaguilera committed:
Header writing ok. Fix some stuff

saantiaguilera committed:
Fix and changed some buffer/write stuff

saantiaguilera committed:
using temporal registers accordingly

saantiaguilera committed:
Indents

saantiaguilera committed:
Using t1 and t2 for buffer and bufLength

aibarbetta committed:
Partial shade loop plus cleanup

saantiaguilera committed:
Me olvide que cambios hice perdon

saantiaguilera committed:
Remove comments. Using align with words

saantiaguilera committed:
Using syscall for io_error because of buffer

saantiaguilera committed:
Indents  (╯°□°)╯︵ ┻━┻

miporto commented on julia-assembly/src/mips32_plot.S:
```diff
@@ -33,13 +35,13 @@ buffer_ini:
 
 .macro empty_buffer
-    lw $t1, buffer_ini # CHANGE INTO PROPER REG
+    lw $t1, buffer_ini # TODO esta chequeado ese label en el lw??
```
No se si ira un load address en vez de un load word, pero seguro hay una manera legal de guardar la posicion de memoria en el registro

miporto committed:
Add push and pop macros

miporto committed:
Typos and labels
Inside macros use arguments with % not with $.
Use labels inside Mips, its allowed ("Labels placed in the body of a
macro definition will not have same name after macro expansion. During
expansion, their name will be followed by "_M#" where # will be a unique
number for each macro expansion.") http://courses.missouristate.edu/KenVollmar/mars/Help/MacrosHelp.html

miporto commented on julia-assembly/src/mips32_plot.S:
```diff
@@ -43,21 +44,18 @@ buffer_ini:
-    sw $data, 0($t1)
-    li $t1, $len
-    addi $t2, $t2, $len
+    blt $t2, BUF_SZ, not_flush_buffer # Store sp location in stack or somewhere so write can jump back
```
Labels are allowed inside macros (see full commit message). No need to store $pc in stack

Manuporto committed:
Delete old comment

miporto committed:
Add some of the shades calculations

miporto committed:
Store results of shade in zr and zi
